compiler/src/dotty/tools/dotc/transform/SpecializeFunction1.scala

```diff
@@ -0,0 +1,164 @@
+package dotty.tools
+package dotc
+package transform
+
+import TreeTransforms.{ MiniPhaseTransform, TransformerInfo }
+import core._
+import Contexts.Context, Types._, Decorators._, Symbols._, DenotTransformers._
+import Denotations._, SymDenotations._, Scopes._, StdNames._, NameOps._, Names._
+import ast.tpd
+
+class SpecializeFunction1 extends MiniPhaseTransform with DenotTransformer {
+  import ast.tpd._
+
+  val phaseName = "specializeFunction1"
+
+  /** Transforms all classes extending `Function1[-T1, +R]` so that
+    * they instead extend the specialized version `JFunction$mp...`
+    */
+  def transform(ref: SingleDenotation)(implicit ctx: Context) = ref match {
+    case ShouldTransformDenot(cref, t1, r, func1) =>
+      transformDenot(cref, t1, r, func1)
+    case _ => ref
+  }
+
+  override def transformTemplate(tree: Template)(implicit ctx: Context, info: TransformerInfo) =
+    tree match {
+      case ShouldTransformTree(func1, t1, r) => transformTree(tree, func1, t1, r)
+      case _ => tree
+    }
+
+  private[this] val functionName = "JFunction1".toTermName
+  private[this] val functionPkg  = "scala.compat.java8.".toTermName
+  private[this] var Function1: Symbol = _
+  private[this] var argTypes: Set[Symbol] = _
+  private[this] var returnTypes: Set[Symbol] = _
+  private[this] var blacklisted: Set[Symbol] = _
+
+  /** Do setup of `argTypes` and `returnTypes` */
+  override def prepareForUnit(tree: Tree)(implicit ctx: Context) = {
+    argTypes = Set(defn.DoubleClass,
+                   defn.FloatClass,
+                   defn.IntClass,
+                   defn.LongClass,
+                   defn.UnitClass,
+                   defn.BooleanClass)
+
+    returnTypes = Set(defn.DoubleClass,
+                      defn.FloatClass,
+                      defn.IntClass,
+                      defn.LongClass)
+
+    Function1 = ctx.requiredClass("scala.Function1")
+
+    blacklisted = Set(
+      "scala.compat.java8.JFunction1",
+      "scala.runtime.AbstractFunction1",
```

Comment:
why do you need to blacklist those 3? As those functions don't know the actual type parameters of parent class, they shoundn't be specialized.